metagraph/src/common/vectors/vector_algorithm.hpp

```diff
@@ -109,6 +110,184 @@ inline void unset_bit(uint64_t *v,
+    } else if ((bit_pos & 0x7F) + width <= 128) {
+        // element fits in an aligned double word
+        __uint128_t *word = &reinterpret_cast<__uint128_t*>(vector.data())[bit_pos >> 7];
+        uint8_t shift = bit_pos & 0x7F;
+        return (__sync_fetch_and_add(word, __uint128_t(count) << shift) >> shift) & mask;
+    } else {
+        uint8_t shift = bit_pos & 0x7;
+        uint8_t *word = &reinterpret_cast<uint8_t*>(vector.data())[bit_pos >> 3];
+        if (shift + width <= 8) {
+            // read from a byte
+            return (__atomic_fetch_add(word, count << shift, memorder) >> shift) & mask;
+        } else if (shift + width <= 16) {
+            // unaligned read from two bytes
+            return (__atomic_fetch_add((uint16_t*)word, count << shift, memorder) >> shift) & mask;
+        } else if (shift + width <= 32) {
+            // unaligned read from four bytes
+            return (__atomic_fetch_add((uint32_t*)word, count << shift, memorder) >> shift) & mask;
+        } else if (shift + width <= 64) {
+            // unaligned read from eight bytes
+            return (__atomic_fetch_add((uint64_t*)word, count << shift, memorder) >> shift) & mask;
+        } else {
+            assert(false);
+        }
+    }
+
+    return 0;
+}
+
+inline uint64_t atomic_exchange(sdsl::int_vector<> &vector, uint64_t i, uint64_t val,
+                                std::mutex &backup_mutex,
+                                int memorder = __ATOMIC_SEQ_CST) {
+    size_t width = vector.width();
+    size_t bit_pos = i * width;
+    uint64_t mask = (1llu << width) - 1;
+    if (width + 7 > 64) {
+        // there is no way to reliably modify without a mutex
+        std::lock_guard<std::mutex> lock(backup_mutex);
+        uint64_t old_val = vector[i];
+        vector[i] = val;
+        return old_val;
+    } else if ((bit_pos & 0x3F) + width <= 64) {
+        // element fits in an aligned word
+        uint64_t *word = &vector.data()[bit_pos >> 6];
+        uint8_t shift = bit_pos & 0x3F;
+        uint64_t desired;
+        uint64_t exp = *word;
+        uint64_t inv_mask = ~(mask << shift);
+        val <<= shift;
+        do {
+            desired = val | (inv_mask & exp);
+        } while (!__atomic_compare_exchange(word, &exp, &desired, true, memorder, __ATOMIC_RELAXED));
+        return (exp >> shift) & mask;
+    } else if ((bit_pos & 0x7F) + width <= 128) {
+        // element fits in an aligned double word
+        __uint128_t *word = &reinterpret_cast<__uint128_t*>(vector.data())[bit_pos >> 7];
+        uint8_t shift = bit_pos & 0x7F;
+        __uint128_t desired;
+        __uint128_t exp;
+        __uint128_t inv_mask = ~(__uint128_t(mask) << shift);
+        __uint128_t big_val = __uint128_t(val) << shift;
+        do {
+            exp = *word;
+            desired = big_val | (inv_mask & exp);
+        } while (!__sync_bool_compare_and_swap(word, exp, desired));
```

Comment:
Add comment here that gcc only generates a `cmpxchg16b` instruction when older `__sync` instructions are used...